File: net/pdfjet/FontStream2.cs
/**
 *  FontStream2.cs
 *
Copyright 2020 Innovatics Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
using System;
using System.IO;
using System.Text;
using System.Collections.Generic;


namespace PDFjet.NET {
class FontStream2 {

    internal static void Register(List<PDFobj> objects, Font font, Stream inputStream) {
        FontStream1.GetFontData(font, inputStream);

        EmbedFontFile(objects, font, inputStream);
        AddFontDescriptorObject(objects, font);
        AddCIDFontDictionaryObject(objects, font);
        AddToUnicodeCMapObject(objects, font);

        // Type0 Font Dictionary
        PDFobj obj = new PDFobj();
        obj.dict.Add("<<");
        obj.dict.Add("/Type");
        obj.dict.Add("/Font");
        obj.dict.Add("/Subtype");
        obj.dict.Add("/Type0");
        obj.dict.Add("/BaseFont");
        obj.dict.Add("/" + font.name);
        obj.dict.Add("/Encoding");
        obj.dict.Add("/Identity-H");
        obj.dict.Add("/DescendantFonts");
        obj.dict.Add("[");
        obj.dict.Add(font.cidFontDictObjNumber.ToString());
        obj.dict.Add("0");
        obj.dict.Add("R");
        obj.dict.Add("]");
        obj.dict.Add("/ToUnicode");
        obj.dict.Add(font.toUnicodeCMapObjNumber.ToString());
        obj.dict.Add("0");
        obj.dict.Add("R");
        obj.dict.Add(">>");
        obj.number = objects.Count + 1;
        font.objNumber = obj.number;
        objects.Add(obj);
    }


    private static int AddMetadataObject(List<PDFobj> objects, Font font) {

        StringBuilder sb = new StringBuilder();
        sb.Append("<?xpacket begin='\uFEFF' id=\"W5M0MpCehiHzreSzNTczkc9d\"?>\n");
        sb.Append("<x:xmpmeta xmlns:x=\"adobe:ns:meta/\">\n");
        sb.Append("<rdf:RDF xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n");
        sb.Append("<rdf:Description rdf:about=\"\" xmlns:xmpRights=\"http://ns.adobe.com/xap/1.0/rights/\">\n");
        sb.Append("<xmpRights:UsageTerms>\n");
        sb.Append("<rdf:Alt>\n");
        sb.Append("<rdf:li xml:lang=\"x-default\">\n");
        sb.Append(font.info);
        sb.Append("</rdf:li>\n");
        sb.Append("</rdf:Alt>\n");
        sb.Append("</xmpRights:UsageTerms>\n");
        sb.Append("</rdf:Description>\n");
        sb.Append("</rdf:RDF>\n");
        sb.Append("</x:xmpmeta>\n");
        sb.Append("<?xpacket end=\"w\"?>");

        byte[] xml = (new System.Text.UTF8Encoding()).GetBytes(sb.ToString());

        // This is the metadata object
        PDFobj obj = new PDFobj();
        obj.dict.Add("<<");
        obj.dict.Add("/Type");
        obj.dict.Add("/Metadata");
        obj.dict.Add("/Subtype");
        obj.dict.Add("/XML");
        obj.dict.Add("/Length");
        obj.dict.Add(xml.Length.ToString());
        obj.dict.Add(">>");
        obj.SetStream(xml);
        obj.number = objects.Count + 1;
        objects.Add(obj);

        return obj.number;
    }


    private static void EmbedFontFile(
            List<PDFobj> objects,
            Font font,
            Stream inputStream) {

        int metadataObjNumber = AddMetadataObject(objects, font);

        PDFobj obj = new PDFobj();
        obj.dict.Add("<<");
        obj.dict.Add("/Metadata");
        obj.dict.Add(metadataObjNumber.ToString());
        obj.dict.Add("0");
        obj.dict.Add("R");
        obj.dict.Add("/Filter");
        obj.dict.Add("/FlateDecode");
        obj.dict.Add("/Length");
        obj.dict.Add(font.compressedSize.ToString());
        if (font.cff) {
            obj.dict.Add("/Subtype");
            obj.dict.Add("/CIDFontType0C");
        }
        else {
            obj.dict.Add("/Length1");
            obj.dict.Add(font.uncompressedSize.ToString());
        }
        obj.dict.Add(">>");
        MemoryStream buf2 = new MemoryStream();
        byte[] buf = new byte[4096];
        int len;
        while ((len = inputStream.Read(buf, 0, buf.Length)) > 0) {
            buf2.Write(buf, 0, len);
        }
        inputStream.Close();
        obj.SetStream(buf2.ToArray());
        obj.number = objects.Count + 1;
        objects.Add(obj);
        font.fileObjNumber = obj.number;
    }


    private static void AddFontDescriptorObject(List<PDFobj> objects, Font font) {
        PDFobj obj = new PDFobj();
        obj.dict.Add("<<");
        obj.dict.Add("/Type");
        obj.dict.Add("/FontDescriptor");
        obj.dict.Add("/FontName");
        obj.dict.Add("/" + font.name);
        obj.dict.Add("/FontFile" + (font.cff ? "3" : "2"));
        obj.dict.Add(font.fileObjNumber.ToString());
        obj.dict.Add("0");
        obj.dict.Add("R");
        obj.dict.Add("/Flags");
        obj.dict.Add("32");
        obj.dict.Add("/FontBBox");
        obj.dict.Add("[");
        obj.dict.Add(font.bBoxLLx.ToString());
        obj.dict.Add(font.bBoxLLy.ToString());
        obj.dict.Add(font.bBoxURx.ToString());
        obj.dict.Add(font.bBoxURy.ToString());
        obj.dict.Add("]");
        obj.dict.Add("/Ascent");
        obj.dict.Add(font.fontAscent.ToString());
        obj.dict.Add("/Descent");
        obj.dict.Add(font.fontDescent.ToString());
        obj.dict.Add("/ItalicAngle");
        obj.dict.Add("0");
        obj.dict.Add("/CapHeight");
        obj.dict.Add(font.capHeight.ToString());
        obj.dict.Add("/StemV");
        obj.dict.Add("79");
        obj.dict.Add(">>");
        obj.number = objects.Count + 1;
        objects.Add(obj);
        font.fontDescriptorObjNumber = obj.number;
    }


    private static void AddToUnicodeCMapObject(List<PDFobj> objects, Font font) {

        StringBuilder sb = new StringBuilder();

        sb.Append("/CIDInit /ProcSet findresource begin\n");
        sb.Append("12 dict begin\n");
        sb.Append("begincmap\n");
        sb.Append("/CIDSystemInfo <</Registry (Adobe) /Ordering (Identity) /Supplement 0>> def\n");
        sb.Append("/CMapName /Adobe-Identity def\n");
        sb.Append("/CMapType 2 def\n");

        sb.Append("1 begincodespacerange\n");
        sb.Append("<0000> <FFFF>\n");
        sb.Append("endcodespacerange\n");

        List<String> list = new List<String>();
        StringBuilder buf = new StringBuilder();
        for (int cid = 0; cid <= 0xffff; cid++) {
            int gid = font.unicodeToGID[cid];
            if (gid > 0) {
                buf.Append('<');
                buf.Append(FontStream1.ToHexString(gid));
                buf.Append("> <");
                buf.Append(FontStream1.ToHexString(cid));
                buf.Append(">\n");
                list.Add(buf.ToString());
                buf.Length = 0;
                if (list.Count == 100) {
                    FontStream1.WriteListToBuffer(sb, list);
                }
            }
        }
        if (list.Count > 0) {
            FontStream1.WriteListToBuffer(sb, list);
        }
        sb.Append("endcmap\n");
        sb.Append("CMapName currentdict /CMap defineresource pop\n");
        sb.Append("end\nend");

        PDFobj obj = new PDFobj();
        obj.dict.Add("<<");
        obj.dict.Add("/Length");
        obj.dict.Add(sb.Length.ToString());
        obj.dict.Add(">>");
        obj.SetStream((new System.Text.UTF8Encoding()).GetBytes(sb.ToString()));
        obj.number = objects.Count + 1;
        objects.Add(obj);
        font.toUnicodeCMapObjNumber = obj.number;
    }


    private static void AddCIDFontDictionaryObject(List<PDFobj> objects, Font font) {

        PDFobj obj = new PDFobj();
        obj.dict.Add("<<");
        obj.dict.Add("/Type");
        obj.dict.Add("/Font");
        obj.dict.Add("/Subtype");
        obj.dict.Add("/CIDFontType" + (font.cff ? "0" : "2"));
        obj.dict.Add("/BaseFont");
        obj.dict.Add("/" + font.name);
        obj.dict.Add("/CIDSystemInfo");
        obj.dict.Add("<<");
        obj.dict.Add("/Registry");
        obj.dict.Add("(Adobe)");
        obj.dict.Add("/Ordering");
        obj.dict.Add("(Identity)");
        obj.dict.Add("/Supplement");
        obj.dict.Add("0");
        obj.dict.Add(">>");
        obj.dict.Add("/FontDescriptor");
        obj.dict.Add(font.fontDescriptorObjNumber.ToString());
        obj.dict.Add("0");
        obj.dict.Add("R");

        float k = 1000.0f / Convert.ToSingle(font.unitsPerEm);
        obj.dict.Add("/DW");
        obj.dict.Add(((int) Math.Round(k * font.advanceWidth[0])).ToString());
        obj.dict.Add("/W");
        obj.dict.Add("[");
        obj.dict.Add("0");
        obj.dict.Add("[");
        for (int i = 0; i < font.advanceWidth.Length; i++) {
            obj.dict.Add(((int) Math.Round(k * font.advanceWidth[i])).ToString());
        }
        obj.dict.Add("]");
        obj.dict.Add("]");

        obj.dict.Add("/CIDToGIDMap");
        obj.dict.Add("/Identity");
        obj.dict.Add(">>");
        obj.number = objects.Count + 1;
        objects.Add(obj);
        font.cidFontDictObjNumber = obj.number;
    }

}   // End of FontStream2.cs
}   // End of namespace PDFjet.NET

File: Sources/PDFjet/FontStream2.swift
/**
 *  FontStream2.swift
 *
Copyright 2023 Innovatics Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import Foundation

class FontStream2 {
    enum StreamError: Error {
        case read
        case write
    }

    static func register(
            _ objects: inout [PDFobj],
            _ font: Font,
            _ stream: InputStream) throws {
        stream.open()
        try FontStream1.getFontData(font, stream)
        embedFontFile(&objects, font, stream)
        stream.close()

        addFontDescriptorObject(&objects, font)
        addCIDFontDictionaryObject(&objects, font)
        addToUnicodeCMapObject(&objects, font)

        // Type0 Font Dictionary
        let obj = PDFobj()
        obj.dict.append("<<")
        obj.dict.append("/Type")
        obj.dict.append("/Font")
        obj.dict.append("/Subtype")
        obj.dict.append("/Type0")
        obj.dict.append("/BaseFont")
        obj.dict.append("/" + font.name)
        obj.dict.append("/Encoding")
        obj.dict.append("/Identity-H")
        obj.dict.append("/DescendantFonts")
        obj.dict.append("[")
        obj.dict.append(String(font.cidFontDictObjNumber))
        obj.dict.append("0")
        obj.dict.append("R")
        obj.dict.append("]")
        obj.dict.append("/ToUnicode")
        obj.dict.append(String(font.toUnicodeCMapObjNumber))
        obj.dict.append("0")
        obj.dict.append("R")
        obj.dict.append(">>")
        obj.number = objects.count + 1
        objects.append(obj)

        font.objNumber = obj.number
    }

    private static func addMetadataObject(
            _ objects: inout [PDFobj],
            _ font: Font) -> Int {
        var sb = String()
        sb.append("<?xpacket id=\"W5M0MpCehiHzreSzNTczkc9d\"?>\n")
        sb.append("<x:xmpmeta xmlns:x=\"adobe:ns:meta/\">\n")
        sb.append("<rdf:RDF xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n")
        sb.append("<rdf:Description rdf:about=\"\" xmlns:xmpRights=\"http://ns.adobe.com/xap/1.0/rights/\">\n")
        sb.append("<xmpRights:UsageTerms>\n")
        sb.append("<rdf:Alt>\n")
        sb.append("<rdf:li xml:lang=\"x-default\">\n")
        sb.append(String(font.info.utf8))
        sb.append("</rdf:li>\n")
        sb.append("</rdf:Alt>\n")
        sb.append("</xmpRights:UsageTerms>\n")
        sb.append("</rdf:Description>\n")
        sb.append("</rdf:RDF>\n")
        sb.append("</x:xmpmeta>\n")
        sb.append("<?xpacket end=\"w\"?>")

        var xml = Array(sb.utf8)

        // This is the metadata object
        let obj = PDFobj()
        obj.dict.append("<<")
        obj.dict.append("/Type")
        obj.dict.append("/Metadata")
        obj.dict.append("/Subtype")
        obj.dict.append("/XML")
        obj.dict.append("/Length")
        obj.dict.append(String(xml.count))
        obj.dict.append(">>")
        obj.setStream(&xml)
        obj.number = objects.count + 1
        objects.append(obj)

        return obj.number
    }


    private static func embedFontFile(
            _ objects: inout [PDFobj],
            _ font: Font,
            _ stream: InputStream) {
        let metadataObjNumber = addMetadataObject(&objects, font)

        let obj = PDFobj()
        obj.dict.append("<<")
        obj.dict.append("/Metadata")
        obj.dict.append(String(metadataObjNumber))
        obj.dict.append("0")
        obj.dict.append("R")
        obj.dict.append("/Filter")
        obj.dict.append("/FlateDecode")
        obj.dict.append("/Length")
        obj.dict.append(String(font.compressedSize!))
        if font.cff {
            obj.dict.append("/Subtype")
            obj.dict.append("/CIDFontType0C")
        } else {
            obj.dict.append("/Length1")
            obj.dict.append(String(font.uncompressedSize!))
        }
        obj.dict.append(">>")
        var buffer2 = [UInt8]()
        var buffer1 = [UInt8](repeating: 0, count: 4096)
        while stream.hasBytesAvailable {
            let count = stream.read(&buffer1, maxLength: buffer1.count)
            if count > 0 {
                buffer2.append(contentsOf: buffer1[0..<count])
            }
        }
        obj.setStream(&buffer2)
        obj.number = objects.count + 1
        objects.append(obj)

        font.fileObjNumber = obj.number
    }

    private static func addFontDescriptorObject(_ objects: inout [PDFobj], _ font: Font) {
        let obj = PDFobj()
        obj.dict.append("<<")
        obj.dict.append("/Type")
        obj.dict.append("/FontDescriptor")
        obj.dict.append("/FontName")
        obj.dict.append("/" + font.name)
        obj.dict.append("/FontFile" + (font.cff ? "3" : "2"))
        obj.dict.append(String(font.fileObjNumber))
        obj.dict.append("0")
        obj.dict.append("R")
        obj.dict.append("/Flags")
        obj.dict.append("32")
        obj.dict.append("/FontBBox")
        obj.dict.append("[")
        obj.dict.append(String(font.bBoxLLx))
        obj.dict.append(String(font.bBoxLLy))
        obj.dict.append(String(font.bBoxURx))
        obj.dict.append(String(font.bBoxURy))
        obj.dict.append("]")
        obj.dict.append("/Ascent")
        obj.dict.append(String(font.fontAscent))
        obj.dict.append("/Descent")
        obj.dict.append(String(font.fontDescent))
        obj.dict.append("/ItalicAngle")
        obj.dict.append("0")
        obj.dict.append("/CapHeight")
        obj.dict.append(String(font.capHeight))
        obj.dict.append("/StemV")
        obj.dict.append("79")
        obj.dict.append(">>")
        obj.number = objects.count + 1
        objects.append(obj)

        font.fontDescriptorObjNumber = obj.number
    }

    private static func addToUnicodeCMapObject(
            _ objects: inout [PDFobj],
            _ font: Font) {
        var sb = String()
        sb.append("/CIDInit /ProcSet findresource begin\n")
        sb.append("12 dict begin\n")
        sb.append("begincmap\n")
        sb.append("/CIDSystemInfo <</Registry (Adobe) /Ordering (Identity) /Supplement 0>> def\n")
        sb.append("/CMapName /Adobe-Identity def\n")
        sb.append("/CMapType 2 def\n")

        sb.append("1 begincodespacerange\n")
        sb.append("<0000> <FFFF>\n")
        sb.append("endcodespacerange\n")

        var list = [String]()
        var buf = String()
        for cid in 0...0xffff {
            let gid = font.unicodeToGID![cid]
            if gid > 0 {
                buf.append("<")
                buf.append(toHexString(gid))
                buf.append("> <")
                buf.append(toHexString(Int(cid)))
                buf.append(">\n")
                list.append(buf)
                buf = ""
                if list.count == 100 {
                    writeListToBuffer(&list, &sb)
                }
            }
        }
        if list.count > 0 {
            writeListToBuffer(&list, &sb)
        }
        sb.append("endcmap\n")
        sb.append("CMapName currentdict /CMap defineresource pop\n")
        sb.append("end\nend")

        var stream = Array(sb.utf8)

        let obj = PDFobj()
        obj.dict.append("<<")
        obj.dict.append("/Length")
        obj.dict.append(String(sb.count))
        obj.dict.append(">>")
        obj.setStream(&stream)
        obj.number = objects.count + 1
        objects.append(obj)

        font.toUnicodeCMapObjNumber = obj.number
    }


    private static func addCIDFontDictionaryObject(
            _ objects: inout [PDFobj],
            _ font: Font) {
        let obj = PDFobj()
        obj.dict.append("<<")
        obj.dict.append("/Type")
        obj.dict.append("/Font")
        obj.dict.append("/Subtype")
        obj.dict.append("/CIDFontType" + (font.cff ? "0" : "2"))
        obj.dict.append("/BaseFont")
        obj.dict.append("/" + font.name)
        obj.dict.append("/CIDSystemInfo")
        obj.dict.append("<<")
        obj.dict.append("/Registry")
        obj.dict.append("(Adobe)")
        obj.dict.append("/Ordering")
        obj.dict.append("(Identity)")
        obj.dict.append("/Supplement")
        obj.dict.append("0")
        obj.dict.append(">>")
        obj.dict.append("/FontDescriptor")
        obj.dict.append(String(font.fontDescriptorObjNumber))
        obj.dict.append("0")
        obj.dict.append("R")

        var k: Float = 1.0
        if font.unitsPerEm != 1000 {
            k = Float(1000.0) / Float(font.unitsPerEm)
        }
        obj.dict.append("/DW")
        obj.dict.append(String(round(k * Float(font.advanceWidth![0]))))
        obj.dict.append("/W")
        obj.dict.append("[")
        obj.dict.append("0")
        obj.dict.append("[")
        for i in 0..<font.advanceWidth!.count {
            obj.dict.append(String(round(k * Float(font.advanceWidth![i]))))
        }
        obj.dict.append("]")
        obj.dict.append("]")

        obj.dict.append("/CIDToGIDMap")
        obj.dict.append("/Identity")
        obj.dict.append(">>")
        obj.number = objects.count + 1
        objects.append(obj)

        font.cidFontDictObjNumber = obj.number
    }

    private static func toHexString(_ code: Int) -> String {
        let str = String(code, radix: 16)
        if str.unicodeScalars.count == 1 {
            return "000" + str
        } else if str.unicodeScalars.count == 2 {
            return "00" + str
        } else if str.unicodeScalars.count == 3 {
            return "0" + str
        }
        return str
    }

    private static func writeListToBuffer(
            _ list: inout [String],
            _ sb: inout String) {
        sb.append(String(list.count))
        sb.append(" beginbfchar\n")
        for str in list {
            sb.append(str)
        }
        sb.append("endbfchar\n")
        list.removeAll()
    }

    private static func getUInt16(_ stream: InputStream) throws -> UInt16 {
        var buffer = [UInt8](repeating: 0, count: 2)
        if stream.read(&buffer, maxLength: 2) == 2 {
            var value = UInt16(buffer[0]) << 8
            value |= UInt16(buffer[1])
            return value
        }
        throw StreamError.read
    }

    private static func getInt8(_ stream: InputStream) throws -> Int {
        var buffer = [UInt8](repeating: 0, count: 1)
        if stream.read(&buffer, maxLength: 1) == 1 {
            return Int(buffer[0])
        }
        throw StreamError.read
    }
}   // End of FontStream2.swift
